crates/mdns/src/lib.rs

```diff
@@ -1,172 +1,435 @@
-            Ok(r) => r,
+    pub fn recv(&mut self) -> MulticastDnsResult<Option<(Vec<u8>, SocketAddr)>> {
+        self.clear_buffer();
+
+        match self.recv_socket.recv_from(&mut self.buffer) {
+            Ok((0, _)) => Ok(None),
+            Ok((num_bytes, addr)) => {
+                debug!(
+                    "Received '{}' bytes: {:?}",
+                    num_bytes,
+                    &self.buffer.to_vec()[..num_bytes]
+                );
+                let packet = self.buffer[..num_bytes].to_vec();
+                Ok(Some((packet, addr)))
+            }
             Err(e) => {
                 if e.kind() == std::io::ErrorKind::WouldBlock {
-                    return Ok(None);
+                    Ok(None)
+                } else {
+                    Err(e.into())
+                }
+            }
+        }
+    }
+
+    /// Clean our buffer of bytes from previous messages.
+    pub fn clear_buffer(&mut self) {
+        self.buffer.zeroize();
+    }
+
+    /// Clean our cache by removing the out of live records.
+    pub fn prune_cache(&mut self) {
+        // Get the entry of the dead records to remove them safely afterward
+        // let mut dead_entry_list: Vec<String> = Vec::with_capacity(self.map_record.len());
+        for (_, records) in self.map_record.iter_mut() {
+            let _: Vec<Record> = records.drain_filter(|r| r.ttl == 0).collect();
+        }
+    }
+
+    /// Update the `time to live` of every cached record.
+    pub fn update_ttl(&mut self) {
+        let own_urls = self.own_urls();
+        for (_netid, records) in self.map_record.iter_mut() {
+            for record in records {
+                if !own_urls.contains(&record.url) {
+                    if record.ttl > 0 {
+                        record.ttl -= 1;
+                    }
+                }
+            }
+        }
+    }
+
+    /// mDNS Querier
+    /// One-Shot Multicast DNS Queries
+    /// Not used in our actual mDNS implementation use case.
+    pub fn query(&mut self) -> MulticastDnsResult<()> {
+        if let Some(query_message) = self.build_query_message() {
+            self.broadcast_message(&query_message)?;
+        }
+
+        Ok(())
+    }
+
+    /// A mDNS Responder that listen to the network in order to defend its name and respond to
+    /// queries
+    /// Not used in our actual mDNS implementation use case.
+    fn responder(&mut self) -> MulticastDnsResult<()> {
+        // Process all elements of the UDP socket stack
+        loop {
+            match self.recv() {
+                Ok(Some((packet, sender_addr))) => {
+                    let dmesg = DnsMessage::from_raw(&packet)?;
+
+                    // Here we update our cache with the responses gathered from the network
+                    if dmesg.nb_answers > 0 {
+                        if let Some(new_map_record) = MapRecord::from_dns_message(&dmesg) {
+                            self.update_cache(&new_map_record);
```

Comment:
We need to do some filtering before we track stuff here:
  - Ignore anything that's not class 1 (INET)
  - Ignore anything that's not type 5 (CNAME)
  - Ignore anything that's not our networkid (we only want to discover holochain nodes that are in our network : )